src/tasks/mod.rs

```diff
@@ -382,44 +332,100 @@ pub type ViewSyncTaskTypes =
 /// add the networking task
 /// # Panics
 /// Is unable to panic. This section here is just to satisfy clippy
-pub async fn add_networking_task(
+pub async fn add_network_task<
+    TYPES: NodeType<ConsensusType = SequencingConsensus, SignatureKey = EncodedSignature>,
+    I: NodeImplementation<
+        TYPES,
+        Leaf = SequencingLeaf<TYPES>,
+        ConsensusMessage = SequencingMessage<TYPES, I>,
+    >,
+    PROPOSAL: ProposalType<NodeType = TYPES>,
+    VOTE: VoteType<TYPES>,
+    MEMBERSHIP: Membership<TYPES>,
+    EXCHANGE: ConsensusExchange<
+            TYPES,
+            Message<TYPES, I>,
+            Proposal = PROPOSAL,
+            Vote = VOTE,
+            Membership = MEMBERSHIP,
+        > + Copy
+        + 'static,
+>(
     task_runner: TaskRunner,
-    event_stream: ChannelStream<GlobalEvent>,
-) -> TaskRunner {
-    let networking_state = NetworkingTaskState {};
+    event_stream: ChannelStream<SequencingHotShotEvent<TYPES, I>>,
+    exchange: EXCHANGE,
+) -> TaskRunner
+where
+    EXCHANGE::Networking:
+        CommunicationChannel<TYPES, Message<TYPES, I>, PROPOSAL, VOTE, MEMBERSHIP>,
+{
+    let channel = exchange.network().clone();
+    let message_stream = unbounded().1.into_stream();
+    let network_state: NetworkTaskState<_, _, _, _, _, _> = NetworkTaskState {
+        channel,
+        event_stream: event_stream.clone(),
+        view: ViewNumber::genesis(),
+        phantom: PhantomData,
+    };
     let registry = task_runner.registry.clone();
-    let networking_event_handler = HandleEvent(Arc::new(move |event, state| {
-        async move {
-            if let GlobalEvent::Shutdown = event {
-                (Some(HotShotTaskCompleted::ShutDown), state)
-            } else {
+    let network_message_handler = HandleMessage(Arc::new(
+        move |message,
+              mut state: NetworkTaskState<
+            TYPES,
+            I,
+            PROPOSAL,
+            VOTE,
+            MEMBERSHIP,
+            EXCHANGE::Networking,
+        >| {
+            async move {
+                state.handle_message(message).await;
                 (None, state)
             }
-        }
-        .boxed()
-    }));
+            .boxed()
+        },
+    ));
+    let network_event_handler = HandleEvent(Arc::new(
+        move |event, mut state: NetworkTaskState<_, _, _, _, MEMBERSHIP, _>| {
+            async move {
+                if let SequencingHotShotEvent::Shutdown = event {
+                    (Some(HotShotTaskCompleted::ShutDown), state)
+                } else {
+                    state.handle_event(event, exchange.membership()).await;
+                    (None, state)
```

Comment:
Would it make sense to check the return value in case it returns `true`? (Can that happen?)

Reply:
Actually the return value of `handle_event` is redundant--it returns `true` only if the event is `Shutdown`, which is already checked by the caller of `handle_event`. I'll remove the return type.